### ISSUE TITLE
internationalize the join page

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-docs#842]
Steps to accomplish this are:
1. Copy index.html to index_es.html. index_es.html will have hardcoded spanish text and will use the spanish version of the dynamic config to fill in the dynamic parts of the page
2. Add a language selector that will toggle between index.html and index_es.html

----- Send PR for review. 

After receiving confirmation that this is implemented correctly

3. Use the `jquery_i18n` library to properly internationalize `index.html`. Remove `index_es.html`
https://github.com/wikimedia/jquery.i18n

----- Send PR for review.

**Comment by shankari** [comment]
Pessimistic schedule
-------------
First PR for review: 10pm Pacific Friday (steps 1 and 2 done)
First review done: 11pm Pacific Friday
PR update based on review comments: midnight Sunday PR time
Second review done: Mid-week (27th or 28th) - in the meanwhile, you should be working on the jquery-i18n exploration and understanding
PR update based on second review comments: Dec 30th
Third review + merge + deploy: Jan 3rd

Optimistic schedule
-------------

First PR for review: 10pm Pacific Friday (steps 1 and 2 done)
First review done: 11pm Pacific Friday
PR update based on review comments: midnight Sunday PR time
Second review done: Mid-week (27th or 28th) - in the meanwhile, you should be working on the jquery-i18n exploration and understanding
PR update based on second review comments + i18n: Dec 30th (including step 3)
Third review + merge + deploy: Jan 3rd (including step 3)

**Comment by shankari** [comment]
@dcolondavila  @sebastiandsc And just to make this even more explicit:
- After you submit the first PR **please do not modify it** so that we can at least deploy the hardcoded join page by the first week of Jan.
- While waiting for my review, you should be working on experimenting with the jquery-i18n library. Make those changes in a **separate** branch created from the first one so that I can review in parallel without messing up the hardcoded solution.

**Comment by shankari** [comment]
@dcolondavila @sebastiandsc I have deployed the modified page to staging
- https://openpath-stage.nrel.gov/join/index.html works
- https://openpath-stage.nrel.gov/join/index_es.html works
- switching from english -> spanish via the dropdown works
- switching from spanish -> english via the dropdown **does not work**

Note that https://openpath-stage.nrel.gov/join/index.html?study_config=stage-program# **does work**

- Please debug this issue and fix
- Please also test more extensively on the staging site and see if there are any other issues
  - If so, please fix them
- Submit another PR for review and re-deploy

**Comment by dcolondavila** [comment]
Yes, we just noticed the mistake that we did regarding on why the Spanish to English button transition is not working. We are working on it right now. We will do another PR in 15 minutes.

As for the Spanish translation of Transportation Secure Data Center, we will change this to "Transporte Centro de Datos Seguro de NREL"

**Comment by dcolondavila** [comment]
New PR is up. We updated the code for the button transition and have tested it multiple times and we determined that it is fully functional. Also, we discovered that the join study link is not working. More info about this at the Join Page PR. Do let us know if anything comes up during the next re-deploy. We will tackle it as soon as possible.

**Comment by shankari** [comment]
@dcolondavila I have now pushed the changes to production.
Next steps:
- [ ] test the UPRM production page at https://uprm-civic-openpath.nrel.gov/join carefully
   - in some earlier testing, you indicated that functionality "did not work". I would encourage you to carefully document your testing steps
   - I believe you were trying to test the nrelopenpath:// URL link, but you actually scanned the QR code to join the study. This does not test the link.
   - If you still believe that some functionality does not work as expected, please indicate exactly what you did, what you expected to happen and what actually happened.
   - please read the open source guidelines to understand the best way to report issues remotely
- [ ] file the follow-on issues I listed while approving your PRs
- [ ] work on using the `jquery-i18n` library in index.html
   - note that you will want to support both the `i18n` json *and* the dynamic config, and combine them using message parameters (https://github.com/wikimedia/jquery.i18n#placeholders)
   - make sure to fall back to English (https://github.com/wikimedia/jquery.i18n#fallback)
   - make sure to read through the library documentation **carefully** since it supports a lot of very cool functionality out of the box

Please let me know when you plan to have the first version of the `jquery-i18n` submitted.

**Comment by dcolondavila** [comment]
We plan on submitting a PR by tomorrow Friday 8am pacific time. If for some reason it gets delayed, then it will be uploaded by 3pm pacific time. This next PR will represent the jquery-i18n version of the join page.

**Comment by dcolondavila** [comment]
# UPRM-CIVIC Join Page Performance Evaluation
We just finished testing the UPRM-CIVIC join page. Here we will explain the steps that we utilized to test the UPRM-CIVIC Join Page, results obtained and the conclusion.

# Steps We Performed For Testing The Join Page:
1. First, we open the UPRM-CIVIC join page English version and we immediately tested for the join page responsiveness. We utilized Chrome's built-in performance tester. The results can be seen below.
2. We verified for any missing context, misspelled words and objects that might be out of place.
3. We verified the console logs to see if the join page was reporting any unusual errors.
4. We performed a functionality test. We tested all the buttons, dropdown menu, web scrolling, checked for broken links and missing images.
5. Then, we preceded on verifying the URPM-CIVIC Join Page Spanish version and performed the steps 1-4 again.

# Join Page English Translation Results (When opening the page for the first time):
- Join Page first time opening the web responsiveness results:
![Capture](https://user-images.githubusercontent.com/70666860/210002368-6ac7f086-b9d7-4cc6-9b3c-388811dcbd04.PNG)
Users should have **no** problem in loading and retendering the join page.
- Join Page English Translation has **no** missing input, context, nor any misspelled word.
- Join Page English Translation has **no** warnings or errors displayed on the console.
- Join Page English Translation **passed** our functionality test.

# Join Page Spanish Version Results (English->Spanish):
- Join Page English->Spanish responsiveness results:
![Capture2](https://user-images.githubusercontent.com/70666860/210004914-5a1f7293-c254-4a6c-942e-18801ccb6c65.PNG)
Users should have **no** problem in transitioning from Spanish to English, loading and retendering the join page.
- Join Page Spanish Translation has **no** missing input, context, nor any misspelled word.
- Join Page Spanish Translation has **no** warnings or errors displayed on the console. Also, the language selector button gives no errors when performing the transition.
- Join Page Spanish Translation **passed** our functionality test.

# Join Page English Version Results (Spanish->English):
- Join Page Spanish->English responsiveness results:
![Capture3](https://user-images.githubusercontent.com/70666860/210004799-82862374-8b7e-4d86-9ec2-3344aec730f3.PNG)
User should have **no** problem in transitioning from English to Spanish, loading and retendering the join page.
- Join Page English Translation has **no** missing input, context, nor any misspelled word.
- Join Page English Translation has **no** warnings or errors displayed on the console. The language selector button gives no errors when performing the transition.
- Join Page English Translation **passed** our functionality test.

#Conclusion:
- Overall, the Join Page's responsiveness looks to be fine. With an average load time of 7ms-9ms and a rendering time of  29ms-50ms.
- The language selector dropdown menu works and it performance the transition from English->Spanish and vice versa as intended.
- The Join page has no broken links and all the buttons work.
- Console log is not reporting any errors or issues.
-  We noticed that whenever we transitioned from Spanish to English, the link of the Join Page changes from https://uprm-civic-openpath.nrel.gov/join/ to https://uprm-civic-openpath.nrel.gov/join/index.html?study_config=uprm-civic. Do let us know if you would like us to do something about this. Nothing in the Join Page changes, only the link. Users can still access the original link.

# Note
- We decided to test the Join Page on our phones web browser. We only verified that everything looks organized, is responsive and that the transition from English ->Spanish and vice versa works. Here are some images of what the web looks like my phone.
https://user-images.githubusercontent.com/70666860/210006271-a36e1f38-cd0a-4355-9c33-432bcda9a4ef.jpg
https://user-images.githubusercontent.com/70666860/210006342-6b2a3185-4a58-4d7a-8399-e08051a07021.jpg
https://user-images.githubusercontent.com/70666860/210006354-a84c4d1e-23a0-4068-8802-6fe4e4921dfd.jpg
- Do let us know if you would like for us to do more testing of the Join Page.
- We plan on submitting a new PR for review by tomorrow Friday 8am Pacific. It will be the first version of the jquery-i18n.

**Comment by shankari** [comment]
@dcolondavila very comprehensive testing!

I would also suggest installing the app using the join screen to confirm that the flow is smooth, and to test out the app in Spanish. Note that the app auto-detects the language based on the phone settings, so you might want to have a primary Spanish speaker (whose phone language is set to Spanish) try out the app install.

> We noticed that whenever we transitioned from Spanish to English, the link of the Join Page changes from https://uprm-civic-openpath.nrel.gov/join/ to https://uprm-civic-openpath.nrel.gov/join/index.html?study_config=uprm-civic. Do let us know if you would like us to do something about this. Nothing in the Join Page changes, only the link. Users can still access the original link.

That's fine. Please see my earlier comment:
https://github.com/e-mission/nrel-openpath-join-page/pull/3#discussion_r1058065012

**Comment by dcolondavila** [comment]
We installed the app, scanned the QR code and joined the study. We have not yet tested the app in Spanish. We haven't been able to find the button that allows us to change the language to Spanish😆. But since the app selects the language based on your phones settings, we will change our phones to Spanish and test it asap. 

We will also implement the follow-on issues you listed while approving the PRs. Which are:
- change the app screenshot to a spanish version
- change the phone app to also read the deployment_partner_name from the new location
- change the dashboard screenshot to a spanish version

**Comment by shankari** [comment]
@dcolondavila there is no button that allows the app to change. It only depends on the phone settings. This is primarily to reduce complexity on the smaller form factor of the phone. When you get to making the phone app changes we can revisit that decision.

**Comment by dcolondavila** [comment]
Update on Join Page jquery.i18n version:

We manage to implement jquery.i18n library to the Join Page, and as of this moment, it works. Whenever we click on the desired language it changes everything to Spanish or to English, except the dynamic configs. As of this moment, we are currently trying to fix the button so that the page can re-load and translate all the dynamic configs whenever the user selects a desired language. Below, you will see a demo video of how to Join Page currently looks like with the implementation of jquery.i18n library. Remember, non of the dynamic configs will be appear translated to Spanish, so they will appear in English. We are fixing this. Do let us know if you would like for us to submit a PR of what we have so that you can review it.

https://user-images.githubusercontent.com/70666860/210113144-906db65b-9ff5-4673-a634-87a152d4f008.mp4

**Comment by shankari** [comment]
> Do let us know if you would like for us to submit a PR of what we have so that you can review it.

Yes, please do submit the PR and mark it as draft to indicate that it is not fully done. In general, high level PRs are good to send out periodically.

**Comment by shankari** [comment]
Fixed by https://github.com/e-mission/nrel-openpath-join-page/pull/6
Next step: internationalize the public dashboard
Related issue for public dashboard i18n: https://github.com/e-mission/e-mission-docs/issues/858

**Comment by shankari** [comment]
I just deployed this on staging and it broke immediately
I haven't looked at the error, but I bet its because you are not using the more complex method to determine the study name - you need to handle stage separately because its format is different from production.

Again, see the original code for an example.

**Comment by shankari** [comment]
ah no, it is because you haven't changed the Dockerfile to copy the files in.
For production builds, we will not load from a file system
You need to also update https://github.com/e-mission/nrel-openpath-join-page/blob/main/frontend/Dockerfile

**Comment by dcolondavila** [comment]
It will be done. We will try to submit this fix by tomorrow with the dashboard PR.

**Comment by dcolondavila** [comment]
@shankari Ok, so we are having difficulty on knowing how to modify a Dokerfile since none of us have any experience with Docker. Do you have any sources that you could provided us so that we can edit the Dockerfile and fix this specific issue?

**Comment by sebastiandsc** [comment]
@shankari I understand that what needs to be modified are the following lines of code?

`
COPY client /public/client
`
`
COPY *.html /public/
`

**Comment by shankari** [comment]
@sebastiandsc Expanded, not modified

**Comment by sebastiandsc** [comment]
@shankari What do you mean by exapanded? Expand to what path?

**Comment by shankari** [comment]
I mean that you should add another COPY statement not change/remove the existing ones

**Comment by dcolondavila** [comment]
@shankari so something like:
`COPY i18n /public/i18n`

**Comment by shankari** [comment]
yes, you would have to test it of couse

**Comment by shankari** [comment]
Still not working:
https://openpath-stage.nrel.gov/join/

```
Loading messages from: i18n/en.json [jquery.i18n.min.js:1:2137]
XHR GEThttps://openpath-stage.nrel.gov/join/i18n/en.json
[HTTP/2 404 Not Found 199ms]

GET https://openpath-stage.nrel.gov/join/favicon.ico
[HTTP/2 404 Not Found 108ms]

Error in loading messages from i18n/en.json Exception: Not Found
```

@dcolondavila @sebastiandsc I am not sure how this ever worked in your testing

**Comment by shankari** [comment]
@dcolondavila @sebastiandsc at this point, all new production deployments, including 
https://uprm-nicr-openpath.nrel.gov/join/
are broken

**Comment by shankari** [comment]
@dcolondavila @sebastiandsc this was actually my error; I hadn't pulled the changes to the internal repo that we use for deployment. After fixing that, it does work on staging.
Feel free to run any performance tests needed
https://openpath-stage.nrel.gov/join/

**Comment by shankari** [comment]
I'm going to close this issue for now. I do want the list of languages and the default language to be pulled from the dynamic config, but that can wait until later.

**Comment by dcolondavila** [comment]
# UPRM-CIVIC Join Page With i18n Performance Evaluation
We just finished testing the UPRM-CIVIC join page with jquery.i18n implemented. We followed the same steps as last time for this evaluation.

# Steps We Performed For Testing The Join Page:
1. First, we open the UPRM-CIVIC join page i18n version (https://openpath-stage.nrel.gov/join/?study_config=uprm-civic) and we immediately tested for the join page responsiveness. We utilized Chrome's built-in performance tester. The results can be seen below.
2. We verified for any missing context, misspelled words and objects that might be out of place.
3. We verified the console logs to see if the join page was reporting any unusual errors.
4. We performed a functionality test. We tested all the buttons, dropdown menu, web scrolling, checked for broken links and missing images.
5. We stress tested the translation button by changing the language of the website multiple times.

# Join Page i18n Version Results:
![image](https://user-images.githubusercontent.com/70666860/227354032-ccc59db2-b5b7-4842-93a4-2da39fd437d2.png)

# Conclusion:
- The Join Page is fully responsive and working properly. it has an average load time of 7ms and a rendering time of 18ms. We noticed that there is a decrease in load time by 2ms - 5ms and in rendering time by 7ms in comparison with the Join Page hardcoded version. This is a big improvement.
- The language selector button works as intended.
- The Join page has no broken links and all the buttons work.
- Console log is not reporting any errors or issues.
- We noticed that whenever we transitioned from Spanish to English, the link of the Join Page changes from https://uprm-civic-openpath.nrel.gov/join/ to https://uprm-civic-openpath.nrel.gov/join/index.html?study_config=uprm-civic. Do let us know if you would like us to do something about this. Nothing in the Join Page changes, only the link. Users can still access the original link.
